key/dashboard/widgets.py

```diff
@@ -1,108 +1,124 @@
-# %% ../../04_widgets.ipynb 70
+    global selected_df
+    selected_df = original_df[(original_df['Year'] >= change['new'][0])
+                              & (original_df['Year'] <= change['new'][1])]
+    # WARNING: You could have thought to use:
+    #  selected_df = original_df[(original_df['Year'] >= year_range.value[0])
+    #                             & (original_df['Year'] <= year_range.value[1])]
+    # but for many manual interactions, such as a click and drag slider, a high
+    # overturn of quickly changing widget values could silently cause
+    # unexpected behavior in changing values, so that syntax should be avoided
+
+# %% ../02b_widgets.ipynb 46
+year_range.observe(on_range_change, 'value') # year_range.observe()
+
+# %% ../02b_widgets.ipynb 62
+from ipydatagrid import DataGrid
+
+# Define a DataGrid widget containing our Pandas dataframe setting column and row sizes
+selected_data_grid = DataGrid(selected_df, header_visibility="column", auto_fit_columns=True)
+
+# %% ../02b_widgets.ipynb 65
+# Create a handler function that will display the selected dataframe
+def update_selected_datagrid(change):
+    selected_data_grid.data = selected_df
+
+# Attach the handler function to the year_range widget
+year_range.observe(update_selected_datagrid, 'value')
+
+# %% ../02b_widgets.ipynb 71
+window_size = widgets.IntSlider(description = 'Window Size', value=20, min=2, max=100,
+                                style={'description_width': 'initial'})
+
+# %% ../02b_widgets.ipynb 74
+poly_order = widgets.BoundedIntText(description = 'Poly Order', min=1, value=3, max=10,
+                                    style={'description_width': 'initial'})
+
+# %% ../02b_widgets.ipynb 83
 def on_poly_order_change(change):
-    global original_data, selected
-    original_data['Savitzky-Golay'] = savgol_filter(original_data['Temperature'], window_size.value, change['new']) # original_data['Savitzky-Golay'] = savgol_filter(original_data['Temperature'], window_size, poly_order)
-    selected = original_data[(original_data['Year'] >= year_range.value[0]) & (original_data['Year'] <= year_range.value[1])]
+    global original_df, selected_df
+    # catch the change in the poly_order widget value and update the original_df
+    original_df['Smoothed Data'] = savgol_filter(original_df['Temperature'], window_size.value, change['new']).round(decimals=3) # original_df['Smoothed Data'] = savgol_filter(original_df['Temperature'], window_size, poly_order).round(decimal=3)
+    selected_df = original_df[(original_df['Year'] >= year_range.value[0]) & (original_df['Year'] <= year_range.value[1])]
 
-# %% ../../04_widgets.ipynb 71
+# %% ../02b_widgets.ipynb 84
 poly_order.observe(on_poly_order_change, 'value')
-poly_order.observe(display_selected_data, 'value')
+window_size.observe(update_selected_datagrid, 'value')
 
-# %% ../../04_widgets.ipynb 76
+# %% ../02b_widgets.ipynb 89
 def on_window_size_change(change):
     global original_data, selected, poly_order
```

Comment:
👀 this one is important -- without it the dashboard doesn't work, because `selected_df` is treated as a local variable. 

```suggestion
    global original_data, selected_df, poly_order
```

Reply:
Oh, when I fixed it in the notebook, I didn't catch it in the answer key. :(. Fixed.